src/metabase/pulse/render.clj

```diff
@@ -336,23 +344,26 @@
         " of "     [:strong {:style (style {:color color-gray-3})} (format-number col-count)]
         " columns."])]))
 
-(defn- render:table
+(s/defn ^:private render:table :- RenderedPulseCard
   [timezone card {:keys [cols rows] :as data}]
-  [:div
-   (render-table (prep-for-html-rendering timezone cols rows nil nil cols-limit))
-   (render-truncation-warning cols-limit (count cols) rows-limit (count rows))])
+  {:attachments nil
+   :content     [:div
+                 (render-table (prep-for-html-rendering timezone cols rows nil nil cols-limit))
+                 (render-truncation-warning cols-limit (count cols) rows-limit (count rows))]})
 
-(defn- render:bar
+(s/defn ^:private render:bar :- RenderedPulseCard
   [timezone card {:keys [cols rows] :as data}]
   (let [max-value (apply max (map second rows))]
-    [:div
-     (render-table (prep-for-html-rendering timezone cols rows second max-value 2))
-     (render-truncation-warning 2 (count cols) rows-limit (count rows))]))
+    {:attachments nil
+     :content     [:div
+                   (render-table (prep-for-html-rendering timezone cols rows second max-value 2))
+                   (render-truncation-warning 2 (count cols) rows-limit (count rows))]}))
 
-(defn- render:scalar
+(s/defn ^:private render:scalar
   [timezone card {:keys [cols rows]}]
-  [:div {:style (style scalar-style)}
-   (h (format-cell timezone (ffirst rows) (first cols)))])
+  {:attachments nil
+   :content     [:div {:style (style scalar-style)}
+                 (h (format-cell timezone (ffirst rows) (first cols)))]})
```

Comment:
Is this supposed to be a `s/defn`? It doesn't look like there's any schema validations :(